src/memcpy_vfp.c

```diff
@@ -0,0 +1,175 @@
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL
+ * THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+ * FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+ * DEALINGS IN THE SOFTWARE.
+ */
+
+#include "driver.h"
+
+#define BLOCK_SIZE  4096
+
+static char tmp_buf0[BLOCK_SIZE] __attribute__((aligned (128)));
+
+static inline void tegra_aligned_block_copy_vfp(void *dst, void *src, int size)
+{
+    asm volatile(
+        "   .fpu vfpv3-d16          \n\t"
+        "   .arch armv7a            \n\t"
+        "0:                         \n\t"
+        "   pld   [%1, #128]        \n\t"
+        "   pld   [%1, #160]        \n\t"
+        "   pld   [%1, #192]        \n\t"
+        "   pld   [%1, #224]        \n\t"
+        "   pld   [%1, #256]        \n\t"
+        "   subs  %2, %2, #128      \n\t"
+        "   vldm  %1!, {d0-d15}     \n\t"
+        "   vstm  %0!, {d0-d15}     \n\t"
+        "   bgt   0b                \n\t"
+        : "+r" (dst), "+r" (src), "+r" (size)
+        :
+        : "cc", "d8", "d9", "d10", "d11", "d12", "d13", "d14", "d15");
+}
+
+static void tegra_do_memcpy_vfp(char *__dst, char *__src, int size,
+                                bool buffering, bool move)
+{
+    int i, dir, block_size = BLOCK_SIZE;
+    char *tmpbuf = tmp_buf0;
+    char *src = (char *)__src;
+    char *dst = __dst;
+
+    if (0) {
+        assert(((unsigned long)src & 127) == 0);
+        assert(((unsigned long)dst & 127) == 0);
+        assert((size & 127) == 0);
+    }
+
+copy:
+    if (size < block_size)
+        block_size = size;
+
+    dir = (dst > src && move) ? -1 : 1;
+
+    if (dir < 0) {
+        src += size - block_size;
+        dst += size - block_size;
+    }
+
+    for (i = 0; i < size / block_size; i++) {
+        if (!buffering)
+            tmpbuf = dst;
+
+        tegra_aligned_block_copy_vfp(tmpbuf, src, block_size);
+
+        if (buffering)
+            tegra_aligned_block_copy_vfp(dst, tmpbuf, block_size);
+
+        src += block_size * dir;
+        dst += block_size * dir;
+    }
+
+    size -= block_size * i;
+    dst = max(dst, __dst);
+    src = max(src, __src);
+
+    if (size)
+        goto copy;
+}
+
+void tegra_memcpy_vfp(void *dst, const void *src, int size, bool src_cached)
+{
+    bool bounce_src = !!((unsigned long)src & 127);
+    bool bounce_dst = !!((unsigned long)dst & 127);
+    int left = size;
+    int block_size;
+    char *csrc = (char *)src;
+    char *cdst = dst;
+    char *__src;
+    char *__dst;
+
+    if (dst == src)
+        return;
```

Comment:
Since performance is the concern here, it would seem better to move this out to the call-sites, as many of them likely can't trigger this... So I would expect this as `assert(dst != src)`. In addition, you probably don't want ranges that overlap either, to avoid read/write order dependencies...

Reply:
Yea, that check could be moved to the header file. `assert` probably is too extreme, but indeed it shouldn't be the case for the current code.

Reply:
Why is assert too extreme? They should only be active in debug-builds...

Reply:
_I think_ it is active by default and should be disabled explicitly by defining `NDEBUG` in the build, we are not defining that.

Reply:
Yeah, you need to opt out. Most sane build systems do this by default, but for autotools [some](https://github.com/mesa3d/mesa/blob/982e012b3ac924dab56b41c5407f722bd2a4c359/configure.ac#L579) [work](https://github.com/mesa3d/mesa/blob/982e012b3ac924dab56b41c5407f722bd2a4c359/configure.ac#L609) is needed to add a switch.

Reply:
Anyway, I get your point. Without a switch like this, perhaps it is too extreme, indeed.

Reply:
Thanks, should be worthwhile to incorporate an --enable-debug configure switch.